packages/backend/server/src/plugins/indexer/service.ts

```diff
@@ -159,6 +167,194 @@ export class IndexerService {
+    const docIds: string[] = [];
+    let cursor: string | undefined;
+    do {
+      const result = await this.search({
+        table: SearchTable.doc,
+        query: {
+          type: SearchQueryType.match,
+          field: 'workspaceId',
+          match: workspaceId,
+        },
+        options: {
+          fields: ['docId'],
+          pagination: {
+            limit: 10000,
+            cursor,
+          },
+        },
+      });
+      docIds.push(...result.nodes.map(node => node.fields.docId[0] as string));
+      cursor = result.nextCursor;
+      this.logger.debug(
+        `get ${result.nodes.length} new / ${docIds.length} total doc ids for workspace ${workspaceId}, nextCursor: ${cursor}`
+      );
+    } while (cursor);
+    return docIds;
+  }
+
+  async syncDoc(
+    workspaceId: string,
+    docId: string,
+    options?: OperationOptions
+  ) {
+    const workspaceSnapshot = await this.models.doc.getSnapshot(
+      workspaceId,
+      workspaceId
+    );
+    if (!workspaceSnapshot) {
+      this.logger.debug(`workspace ${workspaceId} not found`);
+      return;
+    }
+    const docSnapshot = await this.models.doc.getSnapshot(workspaceId, docId);
+    if (!docSnapshot) {
+      this.logger.debug(`doc ${workspaceId}/${docId} not found`);
+      return;
+    }
+    const result = await readAllBlocksFromDocSnapshot(
+      workspaceId,
+      workspaceSnapshot.blob,
+      docId,
+      docSnapshot.blob
+    );
+    if (!result) {
+      this.logger.warn(`parse doc ${workspaceId}/${docId} failed`);
+      return;
+    }
+    await this.write(
+      SearchTable.doc,
+      [
+        {
+          workspaceId,
+          docId,
+          title: result.title,
+          summary: result.summary,
+          // NOTE(@fengmk): journal is not supported yet
+          // journal: result.journal,
+          createdByUserId: docSnapshot.createdBy ?? '',
+          updatedByUserId: docSnapshot.updatedBy ?? '',
+          createdAt: docSnapshot.createdAt,
+          updatedAt: docSnapshot.updatedAt,
+        },
+      ],
+      options
+    );
+    await this.deleteBlocksByDocId(workspaceId, docId, options);
+    await this.write(
+      SearchTable.block,
+      result.blocks.map(block => ({
+        workspaceId,
+        docId,
+        blockId: block.blockId,
+        content: block.content ?? '',
+        flavour: block.flavour,
+        blob: block.blob,
+        refDocId: block.refDocId,
+        ref: block.ref,
+        parentFlavour: block.parentFlavour,
+        parentBlockId: block.parentBlockId,
+        additional: block.additional
+          ? JSON.stringify(block.additional)
+          : undefined,
+        markdownPreview: block.markdownPreview,
+        createdByUserId: docSnapshot.createdBy ?? '',
+        updatedByUserId: docSnapshot.updatedBy ?? '',
+        createdAt: docSnapshot.createdAt,
+        updatedAt: docSnapshot.updatedAt,
+      })),
+      options
+    );
+    this.logger.debug(
```

Comment:
also need `doc.indexer.added` and `doc.indexer.updated` event

Reply:
At present, I don't distinguish between `add` and `update`, they are all directly overridden, so I can only provide `doc.indexer.updated`.

Reply:
reasable, embedding side also treats these two events as the same event, I unified them as updated